Cpp/exercises/numbers (#2187)
exercism/v3: languages/cpp/exercises/concept/numbers/.docs/after.md

### Special integer types

> There are two other important integer types often used in the standard library:

Reviewer: It looks like the `after.md` document adds to what is listed in the `introduction.md` document. As per [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsaftermd), this file should at least contain all information that is introduced in the `introduction.md` document.

- `std::size_t`: unsigned integer type of the result of the `sizeof` operator as well as the `sizeof... operator` and the `alignof` operator (since C++11).
- `std::ptrdiff_t`: is the signed integer type of the result of subtracting two pointers. The bit width of `std::ptrdiff_t` is not less than 17.

### Integer literals

> Integer values can have different written form - "literal" that helps create integer numbers in different notations.

Reviewer: ```suggestion
Integer values can be written using several different notations.
```

Author: Thanks!

### Special integer types

There are two other important integer types often used in the standard library:

- `std::size_t`: unsigned integer type of the result of the `sizeof` operator as well as the `sizeof... operator` and the `alignof` operator (since C++11).
- `std::ptrdiff_t`: is the signed integer type of the result of subtracting two pointers. The bit width of `std::ptrdiff_t` is not less than 17.

### Integer literals

Integer values can have different written form - "literal" that helps create integer numbers in different notations.

- `10` - decimal number
- `010` - octal number
- `0xa` - hexadecimal number
- `0b001` - binary number

### Digit separators

Long numbers in C++ can be separated for readability using `'` (single quotation mark). For example, `10000` is equivalent `10'000`.

### Char types as integers

> Char types `char` `unsigned char`, `wchar_t`, `char16_t` and `char32_t` have the same memory layout and in some cases can be treated as integers.

Author: Indeed, your point is correct. I'll remove the section.

Answering 2, I have never seen `signed char` used anywhere, but `char` used for storing 'small' integers I have seen a lot.

Answering 3, this is also connected to using `char` as numeric type and doing arithmetic on it. It works because `char` is an integer type but it is not obvious in the semantic of character.

### Numeric limits

Compile-time information about a numeric type is available using meta-functions from `<limits>`.

### Conversions

When converting between numeric types, there are two types of numeric conversions:

1. Implicit conversions: generated by the compiler, no additional syntax is required, data maybe be lost: e.g. `double` -> `int`.
2. Explicit conversions: data could be lost and additional syntax in the form of a _cast_ is required.

### Complex numbers

Complex numbers are represented with [`std::complex`][complex] standard module.

### NULL

[NULL][null] is a constant expression of integer type that evaluates to zero.

> 

Author: Constant `nullptr` is of type `std::nullptr_t`, while `NULL` is integer.

Do you think that there should be a mentioning about that in this context?

> 

Reviewer: I would be hesitant to introduce pointers here. That seems like they should be introduced in a separate exercise.

> 

Reviewer 2: I would not introduce a C++ programmer to `NULL`, given it is part of the backwards-compat part of the language.
I think we should strive to first introduce the language as it is today, ignoring the legacy stuff, and then to provide an advanced exercise introducing old stuff so people can read legacy code and understand its behavior.
Besides, `#define NULL nullptr` is a valid C++ implementation of the `NULL` macro (see https://eel.is/c++draft/conv.ptr#1.sentence-1)

> 

Author: OK, will remove.

### `std::is_arithmetic`

> If the type is arithmetic can be done with a meta-function [`std::is_arithmetic`][is_arithmetic].

Author: I agree, this topic is too advance in the context of this exercise. I will remove it.
